sources/api/bootstrap-containers/README.md

```diff
@@ -0,0 +1,75 @@
+# bootstrap-containers
+
+Current version: 0.1.0
+
+## Bootstrap containers
+
+Bootstrap containers are host containers that can be used to setup the host
+before services like the ecs-agent, kubelet and docker start.
+
+This type of containers are very similar to normal host containers, they come with
+persistent storage and with optional user data. Unlike normal host containers, bootstrap
+containers can't be treated as `superpowered`. However, these containers have access to
+the underlying root filesystem through `/.bottlerocket/rootfs`.
+
+Bootstrap containers have three different modes:
+
+* `always`: with this setting, the container is executed on every boot.
+* `off`: blocks the execution of the container
+* `once`: with this setting, the container only runs on the first boot. Upon completion,
+the container marks itself as `off`. If you need to execute the containers after the first
+boot, you can mark it as `once` and reboot.
+
+Bootstrap containers are executed in parallel during the execution of the configured target,
+without following a deterministic order. Therefore, bootstrap containers must be independent
+from each other. Bootstrap containers have the `essential` flag (which defaults to `false`),
+which can be used to fail the boot process if the underlying container task exits with a
+non-zero status code.
+
+Bootstrap containers are created in the `bootstrap` containerd's namespace to prevent
+collisions with other host containers, therefore you can have use the name for both container
+types.
+
+## Examples
+Given a bootstrap container with the following configuration:
+
+```toml
+[settings.bootstrap-containers.bear]
+source="<SOURCE>"
+mode="once"
+user-data="ypXCt82h4bSlwrfKlA=="
+```
+
+Where `<SOURCE>`, is the url of an image with the following definition:
+
+```Dockerfile
+FROM alpine
+ADD bootstrap-script /
+RUN chmod +x /bootstrap-script
+ENTRYPOINT ["sh", "bootstrap-script"]
+```
+
+And `bootstrap-script` as:
+
+```sh
+#!/bin/sh
+BEAR_DIR=/.bottlerocket/bootstrap-containers/bear
+VAR_DIR=/.bottlerocket/rootfs/var
+MY_DIR=$VAR_DIR/lib/my_directory
+cat $BEAR_DIR/user-data
+mkdir $MY_DIR
+chmod -R o+r $MY_DIR
+chown -R 1000:1000 $MY_DIR
+sleep 120
```

Comment:
I suggest we take advantage of [the `$prefix/current` directory which has the user-data](https://github.com/bottlerocket-os/bottlerocket/pull/1416) in this example. We should encourage users to develop their own containers against that path instead of hard coding expected `bootstrap-containers` names within scripts:

```suggestion
#!/usr/bin/env sh
# We'll read some data to be written out from given user-data.
USER_DATA_DIR=/.bottlerocket/bootstrap-containers/current
# This is the in-container view of where the host's `/var` can be accessed.
HOST_VAR_DIR=/.bottlerocket/rootfs/var
# The directory that'll be created by this bootstrap container
MY_HOST_DIR=$VAR_DIR/lib/my_directory

# Create it!
mkdir -p "$MY_HOST_DIR"

# Write the user-data to stdout (to the journal) and to our new path:
tee /dev/stdout "$MY_HOST_DIR/bear.txt" < "$USER_DATA_DIR/user-data" 

# The bootstrap container can set the permissions which are seen by the host:
chmod -R o+r "$MY_HOST_DIR"
chown -R 1000:1000 "$MY_HOST_DIR"

# Bootstrap containers *must* finish before boot continues.
# 
# With this, the boot process will be delayed 120 seconds. You can check the
# status of `preconfigured.target` and `bootstrap-containers@bear` to see
# that this sleep kept the system from starting up the apiserver.
#
# From the admin container:
#
# systemctl status preconfigured.target bootstrap-containers@bear
sleep 120
```

Reply:
Fixed